### PR TITLE
[Iluvatar] Support tensor parallel heterogeneous training

### PR DESCRIPTION
**Pull request by njuerect**
This PR adds support for training LLM on heterogeneous devices with different tensor parallel sizes.

### CODE REVIEW

File: flagscale/train/hetero/train_llama.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (zhaoyinglia) [inline flagscale/train/hetero/train_llama.py:1]: Do not copy a `train_llama.py`, which should be unified training entrance. You can put the functional changes that must be rewritten in hetero folder.

File: flagscale/train/hetero/training.py [hunk 2]
```diff
@@ -0,0 +1,680 @@
+# Copyright (c) 2022, NVIDIA CORPORATION. All rights reserved.
```
Comment (zhaoyinglia) [inline flagscale/train/hetero/training.py:1]: Same as the `train_llama.py` above.

File: flagscale/train/hetero/train_llama.py (hunk 1 above)
Comment (aoyulong) [reply flagscale/train/hetero/train_llama.py:1]: Great idea! This will definitely be improved in the next pull request.

**Review by aoyulong — APPROVED** [review]
LGTM

**Pull request merged**